src/road_gui.cpp

```diff
@@ -304,13 +304,16 @@ struct BuildRoadToolbarWindow : Window {
 	{
 		if (!gui_scope) return;
 
-		if (_game_mode != GM_EDITOR && !CanBuildVehicleInfrastructure(VEH_ROAD, GetRoadTramType(this->roadtype))) delete this;
 		bool can_build = _game_mode != GM_EDITOR;
-		this->SetWidgetsDisabledState(!can_build,
-				WID_ROT_DEPOT,
-				WID_ROT_BUS_STATION,
-				WID_ROT_TRUCK_STATION,
-				WIDGET_LIST_END);
+		if (can_build && !CanBuildVehicleInfrastructure(VEH_ROAD, GetRoadTramType(this->roadtype))) {
+			delete this;
+		} else {
+			this->SetWidgetsDisabledState(!can_build,
+					WID_ROT_DEPOT,
+					WID_ROT_BUS_STATION,
+					WID_ROT_TRUCK_STATION,
+					WIDGET_LIST_END);
+		}
 		if (!can_build) {
```

Comment:
This block should also be executed when the window is destroyed.
Maybe with something like
```suggestion
		if (!can_build || !CanBuildVehicleInfrastructure(VEH_ROAD, GetRoadTramType(this->roadtype))) {
```
 (only tested in my head ;) )

Reply:
I'm not sure. If we close the window, its does not matter whether widgets are disabled or not. And if we reopen the toolbar, the constructor calls this function, so that the widgets will be in the correct state when the new window opens.

Reply:
This part is about closing other build windows (depots, station), it doesn't depend on disabling widgets or not.

Reply:
But if the window is destroyed, it will destroy its child windows too. So the depot and station windows will be deleted too.

Reply:
Ah yes you're right, so I suggested another change to make everything simpler.

Reply:
Hmm and now I'm starting to think that the whole `!can_build` block is actually useless. It was there before c02ef3e because the window was not destroyed, but now it only deletes windows that can't be opened anyway as they are disabled in editor.